mutt_mailbox.c

```diff
@@ -212,8 +213,8 @@ int mutt_mailbox_check(struct Mailbox *m_cur, CheckStatsFlags flags)
     if (!m->visible || !m->poll_new_mail)
       continue;
 
-    CheckStatsFlags m_flags = flags;
-    if (!m->first_check_stats_done && c_mail_check_stats)
+    CheckStatsFlags m_flags = MUTT_MAILBOX_CHECK_NO_FLAGS;
```

Comment:
This means that we don't pass the existing flags through anymore. Won't that mean that `MUTT_MAILBOX_CHECK_IMMEDIATE` no longer gets passed through to the imap code?

Reply:
Interesting.. perhaps we only want to remove the `MUTT_MAILBOX_CHECK_FORCE_STATS` bit from the flags here?

Reply:
Would this work? (diff against the vchimishuk:slow-mbox branch)

```diff
diff --git a/mutt_mailbox.c b/mutt_mailbox.c
index 48f7d6003..81fd0db4a 100644
--- a/mutt_mailbox.c
+++ b/mutt_mailbox.c
@@ -213,10 +213,10 @@ int mutt_mailbox_check(struct Mailbox *m_cur, CheckStatsFlags flags)
     if (!m->visible || !m->poll_new_mail)
       continue;
 
-    CheckStatsFlags m_flags = MUTT_MAILBOX_CHECK_NO_FLAGS;
-    if (force_stats || (!m->first_check_stats_done && c_mail_check_stats))
+    CheckStatsFlags m_flags = flags;
+    if (!(force_stats || (!m->first_check_stats_done && c_mail_check_stats)))
     {
-      m_flags |= MUTT_MAILBOX_CHECK_FORCE_STATS;
+      m_flags &= ~MUTT_MAILBOX_CHECK_FORCE_STATS;
     }
     mailbox_check(m_cur, m, &st_cur, m_flags);
     if (m->has_new)
```

Reply:
This function is a mess. Now it feels weird that we set `force_stats` based on some conditions and then later check it *again* plus some other condition to remove the flag. Can we set `force_stats` in a way that later we just need:
```
if (!force_stats)
  m_flags &= ~MUTT_MAILBOX_CHECK_FORCE_STATS;
```

Reply:
`m_flags` is per mailbox

Reply:
ah right, the `if` statement is still giving me a headache but I think it should work.

Reply:
@gahr, `m_flags &= ~MUTT_MAILBOX_CHECK_FORCE_STATS;` doesn't fix the issue because `flags` has `MUTT_MAILBOX_CHECK_FORCE` flag set not `MUTT_MAILBOX_CHECK_FORCE_STATS`.

I've verified and `m_flags &= ~MUTT_MAILBOX_CHECK_FORCE` does fix the issue.

Reply:
Now I'm confused. So here's the logic in this PR:

1. if flag has `MUTT_MAILBOX_CHECK_FORCE`, we call `mutt_update_num_postponed`
https://github.com/neomutt/neomutt/blob/4309b939c88f410aa1cddf6c04abd04adf89f128/mutt_mailbox.c#L174-L175
ok, I think this is fine and not under debate
1. if no flag is set and `c_mail_check` hasn't elapsed, don't do anything
https://github.com/neomutt/neomutt/blob/4309b939c88f410aa1cddf6c04abd04adf89f128/mutt_mailbox.c#L181-L183
ok, I think this is also fine and not under debate
3. then, we *always* disable `MUTT_MAILBOX_CHECK_FORCE` altogether, and only set `MUTT_MAILBOX_CHECK_FORCE_STATS` if needed
https://github.com/neomutt/neomutt/blob/4309b939c88f410aa1cddf6c04abd04adf89f128/mutt_mailbox.c#L216-L220
where *if needed* is driven by whether `MUTT_MAILBOX_CHECK_FORCE_STATS` was originally part of the flags or whether the `c_mail_check_stats_interval` time has elapsed
https://github.com/neomutt/neomutt/blob/4309b939c88f410aa1cddf6c04abd04adf89f128/mutt_mailbox.c#L185-L191

My confusion is about `MUTT_MAILBOX_CHECK_FORCE`. It looks like now we only use it to call `mutt_update_num_postponed`. It also looks like no backend explicitly checks for `MUTT_MAILBOX_CHECK_FORCE`. There's only mbox but it treats CHECK and CHECK_STATS just the same:
https://github.com/neomutt/neomutt/blob/19653add8a9b372fd078aa98662b4d2c989445a7/mbox/mbox.c#L1691

So, two questions:

1. why removing `MUTT_MAILBOX_CHECK_FORCE` from the flags help?
2. so, should we rename `MUTT_MAILBOX_CHECK_FORCE` to `MUTT_MAILBOX_CHECK_NUM_POSTPONED`?

I'm sure I'm missing something, but I can't pinpoint what it is.

Reply:
> why removing MUTT_MAILBOX_CHECK_FORCE from the flags help?

Because `MUTT_MAILBOX_CHECK_FORCE` is passed into `mutt_mailbox_check()` from `dlg_index()`
https://github.com/neomutt/neomutt/blob/4309b939c88f410aa1cddf6c04abd04adf89f128/index/dlg_index.c#L1103 So, it is `MUTT_MAILBOX_CHECK_FORCE` who triggers the check.

In the original code (changed by https://github.com/neomutt/neomutt/commit/6bad3733ee2d59d04db522d396e4b999b28a2811) `check_stats` flag (passed to `mailbox_check()`) were set only if `force & MUTT_MAILBOX_CHECK_FORCE_STATS`. Now it passes `flags`  as is because of the next line.
```
CheckStatsFlags m_flags = flags;
```
Which makes `MUTT_MAILBOX_CHECK_FORCE` proxied. Which was not true originally. 


> so, should we rename MUTT_MAILBOX_CHECK_FORCE to MUTT_MAILBOX_CHECK_NUM_POSTPONED?

Sorry, I'm not sure I can answer this question. I still has a limited understanding of those constants. However, for me it looks like `MUTT_MAILBOX_CHECK_FORCE` has much wider usage than simply check number of postponed.

Reply:
What do you mean "the wider usage"?

Reply:
Ah! Maybe I *do* get it and this is specifically an issue for `mbox`.

Do we achieve the same result if change this
https://github.com/neomutt/neomutt/blob/19653add8a9b372fd078aa98662b4d2c989445a7/mbox/mbox.c#L1691

into `if (flags & MUTT_MAILBOX_CHECK_FORCE_STATS)`?

Reply:
Also, please check if #4357 makes sense to you

Reply:
> Ah! Maybe I do get it and this is specifically an issue for mbox.

Yes. 

> into if (flags & MUTT_MAILBOX_CHECK_FORCE_STATS)?

Yes. This trick also works.

> Also, please check if https://github.com/neomutt/neomutt/pull/4357 makes sense to you

I checked the code and it does make sense to me. Also I've compiled and tested it on my setup and it works just fine.

Reply:
Thanks a lot! Then, if you're ok with that, I'd close this PR and proceed with mine.

Reply:
Sure. Thank you for getting the fix moving into mainline.

Reply:
Thank you for your patience while I tried to understand the issue :)